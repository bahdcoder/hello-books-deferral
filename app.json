{
  "name": "Hello Books",
  "description": "A system that allows efficient, cost-effective management of Books for Libraries."
}
