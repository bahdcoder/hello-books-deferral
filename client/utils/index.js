export const getDateStr = str => {
    const end = str.indexOf('T');
    const rawDate = str.slice(0, end);
    const firstHyphen = rawDate.indexOf('-');
    const lastHyphen = rawDate.lastIndexOf('-');
    const year = rawDate.slice(0, firstHyphen);
    const month = rawDate.slice(firstHyphen + 1, lastHyphen);
    const day = rawDate.slice(lastHyphen + 1);
    const monthStrArr = [
        'Jan',
        'Feb',
        'Mar',
        'Apr',
        'May',
        'Jun',
        'Jul',
        'Aug',
        'Sep',
        'Oct',
        'Nov',
        'Dec'
    ];
    const monthStr = monthStrArr[Number(month) - 1];
    return `${monthStr} ${day} ${year}`;
};
