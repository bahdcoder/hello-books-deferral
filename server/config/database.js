export default () => ({
    development: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    test: {
        client: process.env.TEST_DATABASE_CLIENT,
        connection: process.env.TEST_DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    review: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.BACKUP_DATABASE,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    staging: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    production: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    }
});
