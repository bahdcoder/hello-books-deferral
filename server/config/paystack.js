export default () => ({
    secret: process.env.PAYSTACK_SECRET_KEY,
    public: process.env.PAYSTACK_PUBLIC_KEY
});
