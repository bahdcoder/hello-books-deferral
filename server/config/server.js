export default () => ({
  port: process.env.PORT || 5000,
  env: process.env.NODE_ENV || "development",
  url: process.env.APP_URL || "http://localhost:5000"
});
