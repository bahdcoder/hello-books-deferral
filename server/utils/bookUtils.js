const { validations } = require("indicative");
const { Vanilla } = require("indicative/builds/formatters");
const Validator = require("indicative/builds/validator");
const Book = require("../models/Book");

/* custom user friendly error messages */
const messages = {
  required: "{{ field }} is required to create a book",
  string: "{{ field }} is not a string",
  number: "{{ field }} is not a number",
  different: "{{ field }} must be unique"
};

// Trim white spaces
const sanitizeRules = {
  title: "trim",
  coverType: "trim",
  description: "trim",
  isbn: "trim",
  publisher: "trim",
  year: "trim",
  copiesAvailable: "trim"
};

/* add the unique custom validator to indicative validations object */
validations.different = async (data, field, message, args, get) => {
  const value = get(data, field);
  if (!value) return;
  const column = args[1];
  const row = await Book.query().where(column, value);
  if (row[0]) throw message;
};

const validatorInstance = Validator(
  validations,
  Vanilla
); /* create custom validator */

module.exports = { validatorInstance, messages, sanitizeRules };
