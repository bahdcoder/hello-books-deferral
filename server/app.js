import express from 'express';
import consola from 'consola';
import bodyParser from 'body-parser';
import swagger from 'swagger-ui-express';
import Knex from 'knex';
import { createServer } from 'http';
import { Model } from 'objection';
import swaggerDocument from '@/swagger.json';
import Routes from '@routes/v1';
import config from '@config';

import 'express-jsend';

export const app = express();

export const databaseConnection = Knex(config.database[config.server.env]);

Model.knex(databaseConnection);

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

Routes(app);

app.get('/', (req, res) => {
    res.status(200).jsend({ message: 'Hello Books Deferral' });
});
/**
 *  setup Swagger
 */
app.use(
    '/api-docs',
    swagger.serve,
    swagger.setup(swaggerDocument, { explorer: true })
);

export const server = createServer(app);

export default { app, databaseConnection };
