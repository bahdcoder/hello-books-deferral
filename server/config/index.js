import dotenv from "dotenv";
import auth from "@config/auth";
import server from "@config/server";
import database from "@config/database";

dotenv.config();

export default {
  auth: auth(),
  server: server(),
  database: database()
};
