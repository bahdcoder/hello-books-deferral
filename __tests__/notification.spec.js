import supertest from 'supertest';
import jwt from 'jsonwebtoken';
import config from '@config';
import { app, databaseConnection } from '@server/app';
import { getUser, createUser } from '@tests/utils/helpers';

const server = () => supertest(app);
const notificationRoute = '/api/v1/notifications/';

describe('NOTIFICATION API ENDPOINTS', () => {
    beforeAll(async () => {
        await databaseConnection.migrate.latest();
        await databaseConnection('users').truncate();
        await databaseConnection('authors').truncate();
        await databaseConnection('notifications').truncate();
    });

    afterAll(async () => {
        await databaseConnection('authors').truncate();
        await databaseConnection('notifications').truncate();
    });

    describe('GET IN-APP NOTIFICATION api/v1/notifications/:email', () => {
        it('should not get notifications if email param is invalid', async () => {
            const user = getUser();

            const createdUser = await createUser(user);

            const token = jwt.sign(
                { id: createdUser.id, email: createdUser.email },
                config.auth.secret
            );

            const { status, body } = await server()
                .get(`${notificationRoute}+d`)
                .set('x-access-token', token);

            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should return user notifications with email param', async () => {
            const user = getUser();
            const createdUser = await createUser(user);

            const token = jwt.sign(
                { id: createdUser.id, email: createdUser.email },
                config.auth.secret
            );

            const { status, body } = await server()
                .get(`${notificationRoute}+johndoe@gmail.com`)
                .set('x-access-token', token);

            expect(status).toBe(200);
            expect(body).toMatchSnapshot();
        });
        it('should not get notification if notification id param is invalid', async () => {
            const user = getUser();

            const createdUser = await createUser(user);

            const token = jwt.sign(
                { id: createdUser.id, email: createdUser.email },
                config.auth.secret
            );

            const { status, body } = await server()
                .get(`${notificationRoute}+sample@email.com/e`)
                .set('x-access-token', token);

            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('SAMPLE NOTIFICATION -> should create author with valid inputs and send user notification', async () => {
            const { status, body } = await server()
                .post('/api/v1/authors')
                .send({ id: 10, name: 'john doe' });

            expect(status).toBe(201);
            expect(body).toMatchSnapshot();
        });
    });
});
