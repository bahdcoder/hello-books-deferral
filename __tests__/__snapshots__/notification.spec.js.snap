// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NOTIFICATION API ENDPOINTS GET IN-APP NOTIFICATION api/v1/notifications/:email SAMPLE NOTIFICATION -> should create author with valid inputs and send user notification 1`] = `
Object {
  "data": Object {
    "id": 1,
    "name": "john doe",
  },
  "status": "success",
}
`;

exports[`NOTIFICATION API ENDPOINTS GET IN-APP NOTIFICATION api/v1/notifications/:email should not get notification if notification id param is invalid 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "id",
      "message": "id is expected to be a an integer",
      "validation": "number",
    },
  ],
  "status": "error",
}
`;

exports[`NOTIFICATION API ENDPOINTS GET IN-APP NOTIFICATION api/v1/notifications/:email should not get notifications if email param is invalid 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "email",
      "message": "email must be a valid",
      "validation": "email",
    },
  ],
  "status": "error",
}
`;

exports[`NOTIFICATION API ENDPOINTS GET IN-APP NOTIFICATION api/v1/notifications/:email should return user notifications with email param 1`] = `
Object {
  "data": Array [],
  "status": "success",
}
`;
